### PR TITLE
[torch] Add native parallel associative scan (aten::associative_scan)

### LINKED ISSUE #95408: Parallel Associative Scan
### 🚀 The feature, motivation and pitch

It would be great to have a parallel prefix sum (associative scan) operation in PyTorch, something like [associative_scan](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.associative_scan.html) in JAX or [scan_associative](https://www.tensorflow.org/probability/api_docs/python/tfp/math/scan_associative) in TensorFlow Probability. This operation is key for the parallelization of some algorithms in CRFs, [filtering/smoothing in state space models](https://github.com/EEA-sensors/sequential-parallelization-examples/blob/main/python/temporal-parallelization-bayes-smoothers/parallel_kalman_jax.ipynb), etc.

### Alternatives

I found [this implementation](https://github.com/lxxue/prefix_sum) but the repository is not being actively maintained. I have not checked if it works though. In any case, it would great to have native support in PyTorch.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by SakshamKapoor2911**
### Summary

Fixes #95408

Issue #95408 requests a native parallel associative scan to unblock SSM/Mamba and Bayesian-filtering workloads that currently have to hand-roll a custom `torch.func.associative_scan` HOP every time. This PR adds a first-class dispatcher operator `aten::associative_scan` with:

- CPU and CUDA native kernels for the four pointwise builtin combines (`add`, `mul`, `max`, `min`) and the `linear_recurrence` combine that powers the Mamba-style SSM recurrence `h[t] = a[t] * h[t-1] + b[t]`.
- Autograd for `add` (reuses `cumsum_backward`), `mul` (`cumprod_backward`), and `max`/`min` (plateau subgradient).
- An Inductor lowering to `ir.Scan.create` (Triton `tl.associative_scan`) with fallback to the eager native op where scan codegen is unavailable.
- A Blelloch tree decomposition in `torch._decomp` for backends without scan codegen support and for export.
- Routing from `torch.func.associative_scan`: builtin pointwise combines that trace to a single `aten.add/mul/maximum/minimum` are classified at trace time and dispatched to the native kernels; arbitrary combines keep using the generic HOP path.

Semantics match `jax.lax.associative_scan` for the builtin combines (inclusive scan, NaN-propagating max/min).

### Design deviation from a `Callable` schema

`native_functions.yaml` schemas have no callable type, so the op registers a dispatcher-safe `str combine_mode`:

```yaml
- func: associative_scan(Tensor self, str combine_mode, int dim=0, bool reverse=False) -> Tensor
  dispatch:
    CompositeExplicitAutograd: associative_scan

- func: associative_scan.TensorList(Tensor[] xs, str combine_mode, int dim=0, bool reverse=False) -> Tensor[]
  dispatch:
    CompositeExplicitAutograd: associative_scan
```

The `TensorList` overload is the Mamba SSM recurrence over `(a, b)` pairs (`combine_mode='linear_recurrence'`). Arbitrary user combine functions keep working through the existing `torch.func.associative_scan` HOP.

### Key Files Changed

- `aten/src/ATen/native/AssociativeScanKernel.h` — dispatch stubs.
- `aten/src/ATen/native/AssociativeScan.cpp` — op wrappers (movedim/reverse, meta impls), `native_functions.yaml` registrations.
- `aten/src/ATen/native/AssociativeScanUtils.h` — shared `ScanVec`/combine functors (host + device).
- `aten/src/ATen/native/cpu/AssociativeScanKernel.cpp` — work-efficient 2-pass chunked parallel prefix (upsweep + prefix-of-partials + downsweep) over `at::parallel_for`.
- `aten/src/ATen/native/cuda/AssociativeScanKernel.cu` — hierarchical block scan: warp-inclusive scan via `__shfl_up_sync`, warp-total scan via shared memory, decoupled look-back over chunks.
- `tools/autograd/derivatives.yaml`, `torch/csrc/autograd/FunctionsManual.*` — `associative_scan_backward` (add/mul/max/min).
- `torch/_inductor/lowering.py` — `aten.associative_scan` -> `ir.Scan.create`, fallback to native op when scan codegen is unavailable.
- `torch/_decomp/decompositions.py` — Blelloch tree decomposition.
- `torch/_higher_order_ops/associative_scan.py` — classify builtin combines and dispatch to native op in eager path.
- `test/test_torch.py`, `test/test_autograd.py` — comprehensive operator + autograd tests.
- `benchmarks/associative_scan_bench.py` — correctness + perf harness against eager `cumsum`/`cumprod` and sequential linear-recurrence reference.

### Known Limitations (Initial Version)

- The `TensorList` (`linear_recurrence`) overload has no autograd formula yet; gradient computation for the SSM recurrence is a follow-up.
- `max`/`min` autograd uses the standard subgradient convention (gradient flows to the first element of each constant plateau of the scan output).
- CPU/CUDA only; XPU and MPS are not implemented in this PR.
- Integer dtypes are supported for `add`/`mul` only; `max`/`min` are floating-point (matching jax).

### Test Plan

Local verification (CUDA 12.8, arch 8.6):

- `pytest test/test_torch.py -k associative_scan` (86 passed)
- `pytest test/test_autograd.py -k associative_scan` (22 passed)
- `python benchmarks/associative_scan_bench.py --cpu --cuda` (ALL CHECKS PASSED)
- `lintrunner` clean across all modified files.

Benchmark results (eager reference in ms; lower is better):

| Size | Native Add | Cumsum | Native Mul | Cumprod | LinRec | Seq Ref |
|---|---|---|---|---|---|---|
| CPU 1M x 64 | 793 | 1479 | 792 | 1525 | 1600 | 53118 |
| CUDA 1M x 64 | 9.5 | 297 | 9.0 | 298 | 15.9 | 107185 |

The native kernel is ~1.9x faster than `cumsum`/`cumprod` on CPU, ~31-33x on CUDA, and the `linear_recurrence` (Mamba) scan is ~33x (CPU) / ~6,800x (CUDA) faster than the sequential reference. `torch.compile` routes `aten::associative_scan` to Triton `tl.associative_scan` on CUDA and falls back to the native op on CPU (verified for all modes including the TensorList overload).

> {AI-Assisted}

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes
